pkg/util/xorm/dialect_spanner.go

```diff
@@ -0,0 +1,284 @@
+	_, exists := spannerReservedKeywords[name]
+	return exists
+}
+func (s *spanner) AndStr() string            { return "AND" }
+func (s *spanner) OrStr() string             { return "OR" }
+func (s *spanner) EqStr() string             { return "=" }
+func (s *spanner) RollBackStr() string       { return "ROLL BACK" }
+func (s *spanner) AutoIncrStr() string       { return "" }    // Spanner does not support auto-increment
+func (s *spanner) SupportInsertMany() bool   { return false } // Needs manual transaction batching
+func (s *spanner) SupportEngine() bool       { return false } // No support for engine selection
+func (s *spanner) SupportCharset() bool      { return false } // ...or charsets
+func (s *spanner) SupportDropIfExists() bool { return false } // Drop should be handled differently
+func (s *spanner) IndexOnTable() bool        { return false }
+func (s *spanner) ShowCreateNull() bool      { return false }
+func (s *spanner) Quote(name string) string  { return "`" + name + "`" }
+func (s *spanner) SqlType(col *core.Column) string {
+	switch col.SQLType.Name {
+	case core.Int, core.BigInt:
+		return "INT64"
+	case core.Varchar, core.Text:
+		return "STRING(MAX)"
+	case core.Bool:
+		return "BOOL"
+	case core.Float, core.Double:
+		return "FLOAT64"
+	case core.Bytea:
+		return "BYTES(MAX)"
+	case core.DateTime, core.TimeStamp:
+		return "TIMESTAMP"
+	default:
+		return "STRING(MAX)" // XXX: more types to add
+	}
+}
+
+func (s *spanner) GetColumns(tableName string) ([]string, map[string]*core.Column, error) {
+	query := `SELECT COLUMN_NAME, SPANNER_TYPE, IS_NULLABLE FROM INFORMATION_SCHEMA.COLUMNS WHERE TABLE_NAME = @tableName`
+	rows, err := s.DB().Query(query, map[string]any{"tableName": tableName})
+	if err != nil {
+		return nil, nil, err
+	}
+	defer rows.Close()
+
+	columns := make(map[string]*core.Column)
+	var colNames []string
+
+	for rows.Next() {
+		var name, sqlType, isNullable string
+		if err := rows.Scan(&name, &sqlType, &isNullable); err != nil {
+			return nil, nil, err
+		}
+
+		col := &core.Column{
+			Name:     name,
+			SQLType:  core.SQLType{Name: sqlType},
+			Nullable: isNullable == "YES",
+		}
+		columns[name] = col
+		colNames = append(colNames, name)
+	}
+
+	return colNames, columns, nil
+}
+
+func (s *spanner) CreateTableSql(table *core.Table, tableName, _, charset string) string {
+	sql := "CREATE TABLE " + s.Quote(tableName) + " ("
+
+	for i, col := range table.Columns() {
+		if i > 0 {
+			sql += ", "
+		}
+		sql += s.Quote(col.Name) + " " + s.SqlType(col)
+		if col.IsPrimaryKey {
+			sql += " PRIMARY KEY"
+		}
+	}
+
+	sql += ") PRIMARY KEY (" + strings.Join(table.PrimaryKeys, ",") + ")"
+	return sql
+}
+
+func (s *spanner) CreateIndexSql(tableName string, index *core.Index) string {
+	sql := "CREATE "
+	if index.Type == core.UniqueType {
+		sql += "UNIQUE NULL_FILTERED "
+	}
+	sql += "INDEX " + index.XName(tableName) + " ON " + tableName + " (" + strings.Join(index.Cols, ", ") + ")"
+	return sql
+}
```

Comment:
Question / nit – Could we / should we use `strings.Builder` here?

Reply:
We could, but this isn't on the hot-path, left as is for now.